### ISSUE TITLE
Allow dismissing the CTA of the Idea Hub widget to activate the module

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3360]
The CTA for activating the Idea Hub module that is conditionally displayed as an alternative version of the Idea Hub widget should be enhanced to be dismissible. Once dismissed, it should be persistently dismissed, i.e. the dismissal should be stored on the server.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The CTA to activate the Idea Hub module (originally implemented in #3275) should be enhanced, now actually receiving the "Close" icon button from [this Figma design](https://www.figma.com/file/YB79eKcGOwyUjlWG8mjDZA/Prompt-Engine-%2F-Idea-Hub?node-id=20332%3A5623).
* When this button is clicked, the CTA should be permanently dismissed:
    * In other words, the widget should now no longer be displayed (return `null`) until (if ever) the user actually activates the module (still possible via Settings, as usual).
    * The dismissal should be stored persistently, i.e. in the WordPress database. This should be stored per user and in a data structure which also scales for potential other dismissals, e.g. it could be an array of dismissed notices, following the example of how we store dismissed feature tours.

## Implementation Brief

**Overview**
This ticket will involve a lot of lift and shift functionality as we've already got most of the work in place for feature tours.

At a high level here is what we're doing, with more detail following the overview:

The functionality to dismiss item after the `X` button is pressed will be the `dismissItem` action from our datastore, that will make a call to our rest endpoint from our `REST_Dismissals_Controller.php` and save the dismissed item to the user option in `googlesitekitpersistent_dismissed_items`.

Prior to rendering the component, we make a call to the `isItemDissmissed` selector, which will use the resolver to call the rest endpoint for `dismissed-items` and let us know whether or not we need to render the component.

**PHP**

- Create a new Directory `includes/Core/Dismissals/`

- Add a new file `includes/Core/Dismissals/Dismissed_Items.php` - this will handle the user settings

- Copy the code from `includes/Core/Feature_Tours/Dismissed_Tours.php` changing the following:
  - Class Name `Dismissed_Tours` - `Dismissed_Items`
  - Option Name `googlesitekitpersistent_dismissed_tours` - `googlesitekitpersistent_dismissed_items`
  - Add a new const `const DISMISS_ITEM_PERMANENTLY = 0;` this signifies that an item is permanently dismissed
  - Method Header `public function add( ...$tour_slug ) {` is changed, along with its implementation:
  - `public function add( $item, $expires_in_seconds = DISMISS_ITEM_PERMANENTLY ){};`
    - The time corresponds to the expiration time in seconds from now, so it's `time() + $expires_in_seconds`, if it's `DISMISS_ITEM_PERMANENTLY`, then the item is permanently dismissed.
    -  Items are added to an associative array which will go in the option. The key is the item, and the expiration time is the value. (e.g. `'savedIdeaNotification' => time() + 24 * 60 * 60` or `'ideaHubWidgetCTA' => 0'`)
    - Add another function `public function is_dismissed( $item )`, which will check if an item is dismissed and return a boolean.
    - The function will work as follows:

1. It takes the value of the option, and iterates every key, for every value that is non zero but less than time(), that is for every temporary dismissal that is in the past, it removes it from the array, thus avoiding bloating the option storage.
2. For the item in question, if it is permanently dismissed (when its value is equal to the constant `DISMISS_ITEM_PERMANENTLY`, or if the value is greater than `time()`, then the return value is `true`, otherwise it will be `false`.
3. If the option value has changed, due to pruning expired temporary dismissals, then the new option value is set.
4. The return value is returned.

- Add another function `public function get_dismissed_items()`, which works similar to the above, but returns all items that are currently dismissed, where the value is `DISMISS_ITEM_PERMANENTLY` or in the future (greater than `time()`)
- Add another function `public function delete( $item ){};` which removes the item from the data storage.

- Add a new file `includes/Core/Dismissals/REST_Dismissals_Controller.php` - this will handle the rest routes

- Copy the code from `includes/Core/Feature_Tours/REST_Feature_Tours_Controller.php` changing the following:

  - Class Name `REST_Feature_Tours_Controller` - `REST_Dismissals_Controller`
  - Property `$dismissed_tours` - `$dismissed_items`
  - In `register()` `/core/user/data/dismissed-tours` - `/core/user/data/dismissed-items`
  - In `get_rest_routes()` `/core/user/data/dismissed-tours` - `/core/user/data/dismissed-items`
  - In `get_rest_routes()` `/core/user/data/dismiss-tour` - `/core/user/data/dismiss-item`
    - In the dismiss-item REST_Route, add a second optional parameter, `expiration`, which is set to the const value`DISMISS_ITEM_PERMANENTLY` by default, validated as an integer value, and passed as a second parameter to the `add( $item, $expiration_time)` function. If a value that is less than `time()` is passed, `add()` is not called as its a temporary dismissal that has already expired.

- Add a new file `includes/Core/Dismissals/Dismissals.php` - this will setup the functionality, enabling the user settings / rest routes

- Copy the code from `includes/Core/Feature_Tours/Feature_Tours.php` changing the following:

  - Class Name `Feature_Tours` - `Dismissals`
  - Property `protected $dismissed_tours` - `protected $dismissed_items` 
  - In `__construct` and `register()` functions, use the new properties and classes.

**UI Image**
Use the image `assets/svg/close.svg` and set the [fill](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/fill) attribute to  `#455A64` in colour (copied from figma)
In `assets/js/modules/idea-hub/components/dashboard/DashboardCTA.js`

**JS**
**Data store**

- Create a new file `assets/js/googlesitekit/datastore/user/dismissed-items.js` 

- Copy the code from `assets/js/googlesitekit/datastore/user/feature-tours.js`, changing the following 

  - `fetchGetDismissedToursStore` - fetchGetDismissedItemsStore
  - `controlCallback` will call `return API.get( 'core', 'user', 'dismissed-items', {}, { useCache: false } );` -- matching our php route above
  - `reducerCallback: ( state, dismissedItemSlugs ) => {`

  - `fetchDismissTourStore` - `fetchDismissItemStore`
  - `baseName: 'dismissItem',`
  - `controlCallback: ( { slug } ) => API.set( 'core', 'user', 'dismiss-item', { slug } ),` -- matching our php route above
  - `baseInitialState` only needs `dismissedItems` (initially undefined)

- From the actions, we need:

  - `dismissTour` - `dismissItem` - we don't need the timestamp for cooldown.

- For the baseResolvers/baseSelectors, we need to add:

  - `getDismissedItems` `isItemDismissed` (based on `isTourDismissed`)

**Component**

- In `assets/js/modules/idea-hub/components/dashboard/DashboardCTA/index.js`

  - Define a new constant, `DISMISS_ITEM_IDEA_HUB_SETUP`, this will be the item that we can dismiss.
  - Add the close button, you can reference the code in `assets/js/components/TourTooltip.js` for how we implement a close button in the UI, use the image `assets/svg/close.svg`, with the fill attribute set as described above, referencing the following for some implementation detail
https://github.com/google/site-kit-wp/blob/bb38c02e896924e0c9c32be4db6246cc217122e6/assets/js/components/TourTooltip.js#L86-L91
  - The button handler will call the `dismissItem` action with the `DISMISS_ITEM_IDEA_HUB_SETUP` constant.
  - Make a call to the `isItemDismissed` selector with the same constant, and if the item is already dismissed, `return null` otherwise return the component as it is implemented.

### Test Coverage

* Add PHP tests in: 
`phpunit/integration/Core/Dismissals/DismissalsTest.php`
`phpunit/integration/Core/Dismissals/REST_Dismissals_ControllerTest.php`

* Add JS tests to cover all actions and selectors in:
 `assets/js/googlesitekit/datastore/user/dismissed-items.test.js` 

* Update stories in
`assets/js/modules/idea-hub/components/dashboard/DashboardCTA/DashboardCTA.stories.js`

### Visual Regression Changes

* Update backstop images relating to the `DashboardCTA` component.
 
## QA Brief

* Disconnect the Idea Hub module
* Go to the dashboard and verify that you see the Idea Hub CTA widget
* Click on the X icon in the top right corner and confirm that the Idea Hub CTA widget disappears
* Refresh the dashboard page and confirm that that the Idea Hub CTA widget doesn't re-appear again

To reset dismissed CTA widget, run the following wp-cli command:

```sh-session
wp user meta delete 1 wp_googlesitekitpersistent_dismissed_items
```

## Changelog entry

* Allow Idea Hub CTA to be dismissed.

**Comment by eugene-manuilov** [comment]
> Create a copy of assets/svg/close.svg to assets/svg/close-alternate.svg and make it #455A64 in colour (copied from figma)
In assets/js/modules/idea-hub/components/dashboard/DashboardCTA.js

@ivankruchkoff We don't need to make a copy of that svg image. We should change the fill color using CSS [fill](https://css-tricks.com/almanac/properties/f/fill/) property.

Also could you please format IB to make it easier to read and understand what needs to be done? Please, use a bulleted or numbered list to separate tasks in the IB.

**Comment by ivankruchkoff** [comment]
Updated @eugene-manuilov and back over to you.

**Comment by eugene-manuilov** [comment]
Thanks, @ivankruchkoff. The IB looks okay to me, although I think we can avoid duplicating PHP functionality by extracting common functionality into parent classes and inherit them in concrete implementations. :thinking: I would like to ask @aaemnnosttv  and @felixarntz (or one of you) to take a look at it too and give a thumb up before we can move it to EB.

**Comment by felixarntz** [comment]
@ivankruchkoff @eugene-manuilov Regarding the abstraction aspect, I think it's okay to initially build the PHP classes separately, with admittedly some duplicate code, since that gives us the flexibility we need. If the final result in the PR proves for them to have enough commonalities to abstract out, we can still do it then, or in a follow-up issue which IMO is a bit safer and reduces initial complexity during engineering the new code (since we can focus on the new functionality rather than fitting with existing code).

One important thing I want to raise here: The need for a dismissal REST endpoint is also present in #3359. However it shows that there may be two different scenarios of dismissal - permanent and temporary. I think it makes sense to build a single endpoint that works for both types of dismissals, rather than adding more and more similar things. Arguably even the feature tour dismissals endpoint is a bit overly specific. I would suggest something like the following:

* A single option `googlesitekit_dismissed_items`, which however is not just an array of slugs, but instead a map of slug and expiration pairs. That would allow us to persistently store temporary dismissals too (which could theoretically be handled client-side too but that is not as reliable due to session storage etc being easily wiped).
    * For example, dismissing `savedIdeaNotification` for 1 day adds `'savedIdeaNotification' => time() + 24 * 60 * 60`.
    * For example, dismissing `ideaHubWidgetCTA` permanently adds `'ideaHubWidgetCTA' => 0'`
* The `Dismissed_Items` class should allow to manage this option, including a method to check if a specific "thing" (i.e. slug) is currently dismissed or not (if expiration is 0, it is dismissed, otherwise it's only dismissed if the expiration is later than now).
    * There should also be a method to return all "things" (slugs) that are currently dismissed (needed for REST endpoint).
* The `REST_Dismissals_Controller`'s `dismiss-item` endpoint should not only require the slug of what to dismiss, but also an optional `expiration` (in seconds) argument. By default it can be `0` (no expiration, i.e. permanent dismissal).

This requires a few changes in the IB, but it doesn't look like major changes from what it has now. With this being a 19, my initial assumption would be the above improvements wouldn't increase the estimate much further. But if you have concerns, we could also break the PHP infrastructure for both #3359 and this out into its separate prerequisite issue.

**Comment by ivankruchkoff** [comment]
@felixarntz makes sense, I've gone ahead and updated to add changes to the rest route, the `add()` function and added two new ones `is_dismissed()` and `get_dismissed_items()`, they prune the option of expired temporary dismissals to avoid bloating the option,  along with defining a const for permanently dismissed items to avoid magic numbers.

Back over to you for review, I also added a note for the IB author on #3359 that this ticket covers the rest endpoint.

**Comment by felixarntz** [comment]
@ivankruchkoff IB mostly LGTM, just a few points (mostly about the expiration):

> If the expiration time is less than 1000000000, then it's the number of seconds we add to now, as a relative offset. If it's greater, then it's used as an epoch time for the temporary item dismissal to complete.

We shouldn't need to support this. Let's use a single expected format (seconds from now), to be specified here.

> In the `dismiss-item` REST_Route, add a second optional parameter, `expiration`, which is set to the const value `DISMISS_ITEM_PERMANENTLY` by default, validated as an integer value, and passed as a second parameter to the `add( $item, $expiration_time)` function. If a value that is less than `time()` is passed, `add()` is not called as its a temporary dismissal that has already expired.

Related to the above, let's consistently use "seconds from now" throughout. Maybe it would be more appropriate to name the argument `expires_in` instead of `expiration` - I'll leave that up to you, but we don't need to support passing a timestamp here since that makes it slightly more complicated.

**Comment by ivankruchkoff** [comment]
@felixarntz updated to use `$expires_in_seconds` for relative offsets only, and also to add a `remove()` method that @asvinb was mentioning needing. It could be used to prune the expired items. Back over to you.

**Comment by felixarntz** [comment]
@ivankruchkoff Sounds good! Thanks, IB ✅

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

**Verified:**

Click on the X icon in the top right corner and the Idea Hub CTA widget disappears. 
When the dashboard page is refreshed, the Idea Hub CTA widget doesn't re-appear again. 
Also verified that the X icon appears on all supported browsers, i.e. Firefox, Safari 
Also verified on smaller screens, i.e. mobiles that the widget is displayed, the X icon appears and the widget disappears.

![image](https://user-images.githubusercontent.com/73545194/124938986-0c435980-e009-11eb-860b-8759ce059033.png)
![image](https://user-images.githubusercontent.com/73545194/124938999-0fd6e080-e009-11eb-93ab-2c5d8224261e.png)

**Comment by felixarntz** [comment]
## Approval ⚠️ 

After dismissing the widget, the empty space doesn't get filled as expected - even after a full page refresh. This wasn't explicitly mentioned in the ACs, but it's clearly a UX flaw - permanently having a blank space there. After dismissal, the dashboard should instead look the same way it would without the Idea Hub widget, i.e. the following widgets in the same area shifting up.

**Comment by felixarntz** [comment]
Marking this as closed, follow-up fix to be implemented via #3722.
